AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/Structure.dfy

```diff
@@ -340,6 +356,63 @@ module {:options "/functionSyntax:4" } Structure {
     Success(map i <- encodedEncryptionContext :: i.0.value := i.1.value)
   }
 
+
+  function SelectCustomEncryptionContextAsString(
+    encryptionContext: Types.EncryptionContextString
+  ): (output: Types.EncryptionContextString)
+    ensures forall k <- output
+              ::
+                (|k| > |ENCRYPTION_CONTEXT_PREFIX|)
+                && k[ .. |ENCRYPTION_CONTEXT_PREFIX|] == ENCRYPTION_CONTEXT_PREFIX
+                && (k in encryptionContext)
+                && encryptionContext[k] == output[k]
+    ensures HIERARCHY_VERSION !in output
+  {
+    var customKeys
+      :=
+      set k <- encryptionContext
+          |
+          (|k| > |ENCRYPTION_CONTEXT_PREFIX|)
+          && k[ .. |ENCRYPTION_CONTEXT_PREFIX|] == ENCRYPTION_CONTEXT_PREFIX
+        ::
+          assert |k| >= |ENCRYPTION_CONTEXT_PREFIX|;
+          assert k[ .. |ENCRYPTION_CONTEXT_PREFIX| ] == ENCRYPTION_CONTEXT_PREFIX;
+          assert HIERARCHY_VERSION[ .. |ENCRYPTION_CONTEXT_PREFIX| ] != ENCRYPTION_CONTEXT_PREFIX by
+          {assert HIERARCHY_VERSION[ .. |ENCRYPTION_CONTEXT_PREFIX| ] == "hierarchy-vers"; }
+          assert HIERARCHY_VERSION != k;
+          k;
+    assume {:axiom} HIERARCHY_VERSION !in customKeys;
```

Comment:
@seebees This Axiom is safe, yes? 
With 4.2.0, I could not get Dafny to accept this truth.